### PR TITLE
Add scope hack for compiler

### PR DESCRIPTION
**Pull request by haya14busa**
This change allow us to translate vimlparser.vim to
other language (e.g. Go) which creates scope in "if" block.

### CODE REVIEW

**Comment by mattn** [comment]
I want to to hear another one's opinion. @blueyed @Kuniwak any thought?

**Comment by blueyed** [comment]
Cannot say anything about it, sorry.

File: autoload/vimlparser.vim [hunk 1]
```diff
@@ -3475,12 +3480,13 @@ function! s:ExprParser.parse_identifier()
...
     let node.pos = npos
     let node.value = curly_parts[0].value
+    return node
   else
     let node = s:Node(s:NODE_CURLYNAME)
     let node.pos = npos
     let node.value = curly_parts
-  endif
   return node
+  endif
```
Comment (mattn) [inline autoload/vimlparser.vim]: add indent space into previous line?

**Comment by mattn** [comment]
@haya14busa Ah, I just understand what you want to fix.

File: js/vimlparser.js [hunk 2]
```diff
@@ -2987,9 +2990,9 @@ ExprParser.prototype.parse_expr8 = function() {
                     if (token.type != TOKEN_SQCLOSE) {
                         throw Err(viml_printf("unexpected token: %s", token.value), token.pos);
                     }
+                    var left = node;
```
Comment (mattn) [inline js/vimlparser.js]: Is this your expected?

**Comment by thinca** [comment]
I think that this can not be maintained continuously.
This should be solved in translater program.

File: js/vimlparser.js (hunk 2 above)
Comment (haya14busa) [inline js/vimlparser.js]: yes? Maybe I just don't understand what you asked?

It's exactly translated as I expected.

**Comment by haya14busa** [comment]
Hmm... to be honest... I agree with you.

This change is the most hackly one to create go-vimlparser and it's indeed hard to maintain.
I think it's also true that we can solve it in translater program.
It's going to be difficult work though...

File: js/vimlparser.js (hunk 2 above)
Comment (mattn) [inline js/vimlparser.js]: Maybe, you don't expect `var` here, I guess.

**Comment by mattn** [comment]
As I mentioned in this issue, https://github.com/vim-jp/vim-vimlparser/issues/61#issuecomment-316400235

We can't this problem, I guess. To fix this issue, we must change `let` as below

current
```
var s = 1
```
newer
```
this.s = 1
```
i.e. We can't use `var` since prepending `var` changes behavior of `exists('s')` .

File: js/vimlparser.js (hunk 2 above)
Comment (haya14busa) [inline js/vimlparser.js]: It works because it's "var", not "let".

Ideally, we may want to drop "var" here but it's not this p-r scope.

Comment (mattn) [inline js/vimlparser.js]: Ah, right.  hopefully we will fix this.

**Comment by haya14busa** [comment]
Hmmm... maybe there is miscommunication. 
There are no problem with current vimlparser including python/javascript version with this p-r except maintainability problem. (JavaScript tests are passed.)

If you talk about this problem as  **general** vimltranslator, yes, we have to care about "exists()", but I think as far as the translater in this repository is for translating autoload/vimlparser.vim, we should not use "exist()" in the first place.
(If you take a look at jscompiler.vim or pycompiler.vim, there are lots of workaround.)

It's indeed a problem if  you want to create general Vim script (to JavaScript) translators, but it's not so high priority problem if you don't have plan to create it.

**Comment by mattn** [comment]
@haya14busa I don't think your changes in this PR still have some problem. I just want to say that we can't make perfect compiler since specification of Vim scripts have ambiguous for that `let` have two meaning `let` and `define`. So I'll merge this.

File: js/vimlparser.js (hunk 2 above)
Comment (haya14busa) [inline js/vimlparser.js]: VIm script's "let" ≒ JavaScript "var"

Cool! Easy to translate! Happy variable scope life! 😇 😇 😇

**Comment by haya14busa** [comment]
Oh, thank you for merging this.

I was about to close this p-r and solve the problem by go-vimlparser side.

Let's revert this p-r if there are maintainability problem. Feel free to revert it.

Hopefully we can enjoy benefit, called "static type-check" by adding gocompiler.vim into this repository soom by including this p-r!

**Comment by mattn** [comment]
Okay, I agree with your suggestion. looking forward your p-r.

**Pull request merged**